### ISSUE TITLE
feat: Visual Detection & Recognition - Object Detection, OCR, and Instance Segmentation

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#580]
## Summary

Implement comprehensive visual detection and recognition capabilities including object detection, OCR/text detection, and instance segmentation. These are critical computer vision features that are currently missing from AiDotNet.

## Motivation

- **No existing object detection**: Cannot detect or localize objects in images
- **No OCR/text detection**: Cannot extract text from images or documents
- **No instance segmentation**: Cannot generate pixel-level masks for detected objects
- These are fundamental computer vision tasks required for production applications

### Facade Pattern Compliance

All implementations MUST follow the existing facade design pattern:

1. **Training via PredictionModelBuilder**
   ```csharp
   var result = new PredictionModelBuilder<float, Matrix<float>, Vector<float>>()
       .ConfigurePreprocessing(new ImagePreprocessor<float>())
       .ConfigureModel(new YOLOv8<float>(ModelSize.Medium))
       .ConfigureTrainingPipeline(pipeline => pipeline
           .AddObjectDetectionStage(stage => {
               stage.TrainingData = cocoData;
               stage.Epochs = 100;
           }))
       .Build(images, annotations);
   ```

2. **Inference via PredictionModelResult**
   ```csharp
   var detections = result.Predict(inputImage);
   // or with post-processing
   var results = result.DetectObjects(inputImage, confidence: 0.5, nmsThreshold: 0.4);
   ```

3. **Serving via AiDotNet.Serving**
   - REST endpoints for detection inference
   - Batching support for high-throughput scenarios
   - Model versioning and A/B testing

### Internal Implementation (Hidden from Users)

All backbone networks, detection heads, loss functions, and post-processing should be internal implementation details. Users only interact through the facade.

### 1. Object Detection (~3,500 LOC)

**Model Families (Equal Priority)**

| Family | Models | Description |
|--------|--------|-------------|
| YOLO | YOLOv8, YOLOv9, YOLOv10, YOLOv11 | Real-time single-stage detectors |
| DETR | DETR, DINO, RT-DETR | Transformer-based detection |
| R-CNN | Faster R-CNN, Cascade R-CNN | Two-stage detectors |

**Core Components**

- **Backbone Networks**: CSPDarknet, ResNet, Swin Transformer, EfficientNet
- **Neck Modules**: FPN, PANet, BiFPN
- **Detection Heads**: Anchor-based, anchor-free, transformer decoders
- **Loss Functions**: CIoU, DIoU, GIoU, focal loss, DETR set loss
- **Anchors**: Anchor generation, K-means clustering for custom anchors

**Configuration Options**

```csharp
public class ObjectDetectionOptions<T>
{
    public DetectionArchitecture Architecture { get; set; }
    public ModelSize Size { get; set; }  // Nano, Small, Medium, Large, XLarge
    public int NumClasses { get; set; }
    public float ConfidenceThreshold { get; set; }
    public float NmsThreshold { get; set; }
    public int MaxDetections { get; set; }
    public bool UseMultiScale { get; set; }
}
```

### 2. OCR and Text Detection (~2,500 LOC)

**Both Scene Text AND Document OCR**

| Type | Use Case | Models |
|------|----------|--------|
| Scene Text | Natural images, signs, products | CRAFT, EAST, DBNet |
| Document OCR | Forms, receipts, structured docs | Layout analysis + recognition |

**Core Components**

- **Text Detection**: Localize text regions in images
- **Text Recognition**: CRNN, TrOCR, transformer-based
- **End-to-End**: Combined detection + recognition pipelines
- **Layout Analysis**: Document structure understanding
- **Post-Processing**: Text line grouping, reading order

**Configuration Options**

```csharp
public class OCROptions<T>
{
    public OCRMode Mode { get; set; }  // SceneText, Document, Both
    public string[] SupportedLanguages { get; set; }
    public bool DetectOrientation { get; set; }
    public bool CorrectSkew { get; set; }
    public float TextConfidenceThreshold { get; set; }
    public bool GroupTextLines { get; set; }
}
```

### 3. Instance Segmentation (~2,000 LOC)

**Models**

- Mask R-CNN
- YOLO-Seg (YOLOv8-Seg)
- SOLOv2
- CondInst

**Core Components**

- **Mask Heads**: Parallel mask prediction branches
- **Mask Loss**: Binary cross-entropy, dice loss
- **Mask Post-Processing**: Mask NMS, mask refinement

### 4. Post-Processing Pipeline (~1,200 LOC)

- **NMS Variants**: Standard, Soft-NMS, DIoU-NMS, batched NMS
- **Object Tracking**: SORT, DeepSORT, ByteTrack (video sequences)
- **Visualization**: Bounding boxes, masks, labels, confidence scores

### 5. Pre-trained Weights Support (~800 LOC)

**Facade-compliant weight loading**

```csharp
// Training with pre-trained backbone (hidden complexity)
var builder = new PredictionModelBuilder<float, Matrix<float>, Vector<float>>()
    .ConfigureModel(new YOLOv8<float>(ModelSize.Medium, pretrained: true))
    .ConfigureFineTuning(config => config.FreezeBackbone = true);

// Users do not see weight file handling - it is internal
```

**Supported Sources**
- COCO pre-trained weights (bundled or downloaded)
- ImageNet pre-trained backbones
- Custom checkpoint loading (internal implementation)

## Folder Structure (By Task Type)

```
src/
  ComputerVision/
    Detection/
      ObjectDetection/
        ObjectDetectorBase.cs
        ObjectDetectionOptions.cs
        YOLO/
          YOLOv8.cs
          YOLOv9.cs
          YOLOv10.cs
          YOLOv11.cs
          YOLOBackbone.cs
        DETR/
          DETR.cs
          DINO.cs
          RTDETR.cs
          TransformerDecoder.cs
        RCNN/
          FasterRCNN.cs
          CascadeRCNN.cs
          RPN.cs
        Backbones/
          CSPDarknet.cs
          ResNet.cs
          SwinTransformer.cs
          EfficientNet.cs
        Necks/
          FPN.cs
          PANet.cs
          BiFPN.cs
        Anchors/
          AnchorGenerator.cs
          AnchorMatcher.cs
        Losses/
          CIoULoss.cs
          FocalLoss.cs
          DETRSetLoss.cs
        PostProcessing/
          NMS.cs
          SoftNMS.cs
          DIoUNMS.cs

      TextDetection/
        TextDetectorBase.cs
        CRAFT.cs
        EAST.cs
        DBNet.cs

    Segmentation/
      InstanceSegmentation/
        InstanceSegmenterBase.cs
        InstanceSegmentationOptions.cs
        MaskRCNN.cs
        YOLOSeg.cs
        SOLOv2.cs
        MaskHead.cs
        MaskLoss.cs

    OCR/
      OCRBase.cs
      OCROptions.cs
      Recognition/
        CRNN.cs
        TrOCR.cs
      EndToEnd/
        SceneTextReader.cs
        DocumentReader.cs
      Layout/
        LayoutAnalyzer.cs
        ReadingOrderDetector.cs

    Tracking/
      ObjectTracker.cs
      SORT.cs
      DeepSORT.cs
      ByteTrack.cs

    Visualization/
      DetectionVisualizer.cs
      MaskVisualizer.cs
      OCRVisualizer.cs

    Weights/
      WeightLoader.cs
      WeightConverter.cs
      PretrainedWeightRegistry.cs

  Models/Options/
    ObjectDetectionOptions.cs
    OCROptions.cs
    InstanceSegmentationOptions.cs
    DetectionTrainingPipelineExtensions.cs
```

### PredictionModelBuilder Extensions

```csharp
// Add to TrainingPipelineConfiguration
public static class DetectionTrainingPipelineExtensions
{
    public static TrainingPipelineConfiguration<T, TInput, TOutput> AddObjectDetectionStage<T, TInput, TOutput>(...);
    public static TrainingPipelineConfiguration<T, TInput, TOutput> AddOCRTrainingStage<T, TInput, TOutput>(...);
    public static TrainingPipelineConfiguration<T, TInput, TOutput> AddSegmentationStage<T, TInput, TOutput>(...);
}
```

### PredictionModelResult Extensions

```csharp
// Extension methods for inference
public static class DetectionInferenceExtensions
{
    public static DetectionResult<T> DetectObjects<T>(this PredictionModelResult<T, ...> result, ...);
    public static OCRResult<T> RecognizeText<T>(this PredictionModelResult<T, ...> result, ...);
    public static SegmentationResult<T> SegmentInstances<T>(this PredictionModelResult<T, ...> result, ...);
}
```

### AiDotNet.Serving Endpoints

- `POST /api/detection/objects` - Object detection
- `POST /api/ocr/text` - Text recognition
- `POST /api/segmentation/instances` - Instance segmentation
- `POST /api/detection/batch` - Batch processing

## Performance Modes

**Balanced approach with configurable modes:**

```csharp
public enum PerformanceMode
{
    Speed,      // Nano/Small models, lower resolution
    Balanced,   // Medium models, standard resolution
    Accuracy    // Large models, multi-scale inference
}
```

## Estimated Effort

| Component | Lines of Code | Priority |
|-----------|---------------|----------|
| Object Detection Core | ~2,000 | HIGH |
| YOLO Family | ~800 | HIGH |
| DETR Family | ~600 | HIGH |
| R-CNN Family | ~600 | MEDIUM |
| Backbones and Necks | ~1,000 | HIGH |
| Text Detection | ~800 | HIGH |
| Text Recognition | ~600 | HIGH |
| OCR End-to-End | ~500 | MEDIUM |
| Instance Segmentation | ~1,200 | HIGH |
| Tracking | ~600 | MEDIUM |
| Post-Processing | ~800 | HIGH |
| Visualization | ~400 | LOW |
| Weight Management | ~600 | MEDIUM |
| **Total** | **~10,500** | - |

## Success Criteria

- [ ] Object detection achieves COCO-comparable mAP with YOLO models
- [ ] OCR achieves >95% accuracy on standard benchmarks
- [ ] Instance segmentation produces accurate pixel-level masks
- [ ] All implementations follow facade pattern (users never see internal complexity)
- [ ] Training via PredictionModelBuilder, inference via PredictionModelResult
- [ ] Performance: Real-time inference (>30 FPS) with appropriate model sizes
- [ ] Full integration with AiDotNet.Serving for production deployment

## Related Issues

- #578 - Image Augmentation (dependency for training)
- #392 - CNN Architectures (shared backbone components)
- #387 - Critical Preprocessing Transformers (image preprocessing)

## References

- YOLOv8: https://github.com/ultralytics/ultralytics
- DETR: https://github.com/facebookresearch/detr
- Mask R-CNN: https://arxiv.org/abs/1703.06870
- CRAFT: https://github.com/clovaai/CRAFT-pytorch
- TrOCR: https://arxiv.org/abs/2109.10282
